### ISSUE TITLE
Repeated "interface - Grabbing hotkey" causing high CPU load

### ISSUE THREAD

**Issue by wviechtb** [autokey/autokey#193]
## Classification:

Performance

## Reproducibility:

Always

## Summary

When starting `autokey-gtk -l`, I get lots of repeated `DEBUG - interface - Grabbing hotkey:` messages for the same hotkey. With lots of hotkeys defined, this creates high CPU loads.

## Steps to Reproduce (if applicable)

I just need to start `autokey-gtk -l`.

## Expected Results

Each hotkey should only be grabbed once.

## Actual Results

Lots of repeated `DEBUG - interface - Grabbing hotkey:` messages for the same hotkey.

```
2018-09-23 21:52:58,795 INFO - root - Initialising application
2018-09-23 21:52:58,801 INFO - root - Initialise global hotkeys
2018-09-23 21:52:58,801 INFO - config-manager - Loading config from existing file: /home/wviechtb/.config/autokey/autokey.json
2018-09-23 21:52:58,802 DEBUG - config-manager - Loading folder at '/home/wviechtb/.config/autokey/data/Sample Scripts'
2018-09-23 21:52:58,803 DEBUG - config-manager - Loading folder at '/home/wviechtb/.config/autokey/data/My Phrases'
2018-09-23 21:52:58,804 INFO - config-manager - Configuration changed - rebuilding in-memory structures
2018-09-23 21:52:58,804 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey/data/Sample Scripts
2018-09-23 21:52:58,804 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey/data/My Phrases
2018-09-23 21:52:58,805 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey/data/My Phrases/Addresses
2018-09-23 21:52:58,805 INFO - config-manager - Successfully loaded configuration
2018-09-23 21:52:58,805 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey/data
2018-09-23 21:52:58,805 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey
2018-09-23 21:52:58,805 DEBUG - config-manager - Global settings: {'undoUsingBackspace': True, 'showTrayIcon': True, 'showToolbar': False, 'enableQT4Workaround': False, 'isFirstRun': True, 'interfaceType': 'XRecord', 'hPanePosition': 641, 'triggerItemByInitial': False, 'workAroundApps': '.*VirtualBox.*|krdc.Krdc', 'windowDefaultSize': [2560, 1407], 'menuTakesFocus': False, 'sortByUsageCount': True, 'promptToSave': False, 'columnWidths': [150, 50, 100], 'notificationIcon': 'autokey-status', 'scriptGlobals': {}, 'serviceRunning': True}
2018-09-23 21:52:58,805 INFO - service - Starting service
2018-09-23 21:52:58,829 DEBUG - interface - Modifier masks: {<Key.ALT: '<alt>'>: 8, <Key.SUPER: '<super>'>: 64, <Key.META: '<meta>'>: 8, <Key.ALT_GR: '<alt_gr>'>: 128, <Key.CONTROL: '<ctrl>'>: 4, <Key.SHIFT: '<shift>'>: 1, <Key.HYPER: '<hyper>'>: 64, <Key.NUMLOCK: '<numlock>'>: 16}
2018-09-23 21:52:58,865 DEBUG - interface - Alt-Grid: XK_Alt_R, 65514
2018-09-23 21:52:58,866 DEBUG - interface - <map object at 0x7f44e83ad978>
2018-09-23 21:52:58,866 DEBUG - interface - X Server Keymap
2018-09-23 21:52:58,867 DEBUG - interface - [\]: [(51, 0), (51, 2)]
2018-09-23 21:52:58,867 DEBUG - interface - [|]: [(51, 1), (51, 3), (94, 4), (94, 6)]
2018-09-23 21:52:58,867 DEBUG - interface - [`]: [(49, 0), (49, 2)]
2018-09-23 21:52:58,867 DEBUG - interface - [1]: [(10, 0), (10, 2)]
2018-09-23 21:52:58,867 DEBUG - interface - [2]: [(11, 0), (11, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [3]: [(12, 0), (12, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [4]: [(13, 0), (13, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [5]: [(14, 0), (14, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [6]: [(15, 0), (15, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [7]: [(16, 0), (16, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [8]: [(17, 0), (17, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [9]: [(18, 0), (18, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [0]: [(19, 0), (19, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [-]: [(20, 0), (20, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [=]: [(21, 0), (21, 2)]
2018-09-23 21:52:58,868 DEBUG - interface - [~]: [(49, 1), (49, 3)]
2018-09-23 21:52:58,868 DEBUG - interface - [!]: [(10, 1), (10, 3)]
2018-09-23 21:52:58,868 DEBUG - interface - [@]: [(11, 1), (11, 3)]
2018-09-23 21:52:58,868 DEBUG - interface - [#]: [(12, 1), (12, 3)]
2018-09-23 21:52:58,869 DEBUG - interface - [$]: [(13, 1), (13, 3)]
2018-09-23 21:52:58,869 DEBUG - interface - [%]: [(14, 1), (14, 3)]
2018-09-23 21:52:58,869 DEBUG - interface - [^]: [(15, 1), (15, 3)]
2018-09-23 21:52:58,869 DEBUG - interface - [&]: [(16, 1), (16, 3)]
2018-09-23 21:52:58,869 DEBUG - interface - [*]: [(17, 1), (17, 3)]
2018-09-23 21:52:58,869 DEBUG - interface - [(]: [(187, 0), (18, 1), (187, 2), (18, 3)]
2018-09-23 21:52:58,869 DEBUG - interface - [)]: [(188, 0), (19, 1), (188, 2), (19, 3)]
2018-09-23 21:52:58,869 DEBUG - interface - [q]: [(24, 0), (24, 2)]
2018-09-23 21:52:58,869 DEBUG - interface - [w]: [(25, 0), (25, 2)]
2018-09-23 21:52:58,869 DEBUG - interface - [e]: [(26, 0), (26, 2)]
2018-09-23 21:52:58,869 DEBUG - interface - [r]: [(27, 0), (27, 2)]
2018-09-23 21:52:58,869 DEBUG - interface - [t]: [(28, 0), (28, 2)]
2018-09-23 21:52:58,869 DEBUG - interface - [y]: [(29, 0), (29, 2)]
2018-09-23 21:52:58,869 DEBUG - interface - [u]: [(30, 0), (30, 2)]
2018-09-23 21:52:58,869 DEBUG - interface - [i]: [(31, 0), (31, 2)]
2018-09-23 21:52:58,869 DEBUG - interface - [o]: [(32, 0), (32, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [p]: [(33, 0), (33, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [[]: [(34, 0), (34, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - []]: [(35, 0), (35, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [a]: [(38, 0), (38, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [s]: [(39, 0), (39, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [d]: [(40, 0), (40, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [f]: [(41, 0), (41, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [g]: [(42, 0), (42, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [h]: [(43, 0), (43, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [j]: [(44, 0), (44, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [k]: [(45, 0), (45, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [l]: [(46, 0), (46, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [;]: [(47, 0), (47, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [']: [(48, 0), (48, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [z]: [(52, 0), (52, 2)]
2018-09-23 21:52:58,870 DEBUG - interface - [x]: [(53, 0), (53, 2)]
2018-09-23 21:52:58,871 DEBUG - interface - [c]: [(54, 0), (54, 2)]
2018-09-23 21:52:58,871 DEBUG - interface - [v]: [(55, 0), (55, 2)]
2018-09-23 21:52:58,871 DEBUG - interface - [b]: [(56, 0), (56, 2)]
2018-09-23 21:52:58,873 DEBUG - interface - [n]: [(57, 0), (57, 2)]
2018-09-23 21:52:58,873 DEBUG - interface - [m]: [(58, 0), (58, 2)]
2018-09-23 21:52:58,873 DEBUG - interface - [,]: [(59, 0), (59, 2)]
2018-09-23 21:52:58,874 DEBUG - interface - [.]: [(60, 0), (60, 2)]
2018-09-23 21:52:58,874 DEBUG - interface - [/]: [(61, 0), (61, 2)]
2018-09-23 21:52:58,874 DEBUG - interface - [Q]: [(24, 1), (24, 3)]
2018-09-23 21:52:58,874 DEBUG - interface - [W]: [(25, 1), (25, 3)]
2018-09-23 21:52:58,874 DEBUG - interface - [E]: [(26, 1), (26, 3)]
2018-09-23 21:52:58,874 DEBUG - interface - [R]: [(27, 1), (27, 3)]
2018-09-23 21:52:58,874 DEBUG - interface - [T]: [(28, 1), (28, 3)]
2018-09-23 21:52:58,874 DEBUG - interface - [Y]: [(29, 1), (29, 3)]
2018-09-23 21:52:58,874 DEBUG - interface - [U]: [(30, 1), (30, 3)]
2018-09-23 21:52:58,874 DEBUG - interface - [I]: [(31, 1), (31, 3)]
2018-09-23 21:52:58,874 DEBUG - interface - [O]: [(32, 1), (32, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [P]: [(33, 1), (33, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [{]: [(34, 1), (34, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [}]: [(35, 1), (35, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [A]: [(38, 1), (38, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [S]: [(39, 1), (39, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [D]: [(40, 1), (40, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [F]: [(41, 1), (41, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [G]: [(42, 1), (42, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [H]: [(43, 1), (43, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [J]: [(44, 1), (44, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [K]: [(45, 1), (45, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [L]: [(46, 1), (46, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [:]: [(47, 1), (47, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - ["]: [(48, 1), (48, 3)]
2018-09-23 21:52:58,875 DEBUG - interface - [Z]: [(52, 1), (52, 3)]
2018-09-23 21:52:58,876 DEBUG - interface - [X]: [(53, 1), (53, 3)]
2018-09-23 21:52:58,876 DEBUG - interface - [C]: [(54, 1), (54, 3)]
2018-09-23 21:52:58,876 DEBUG - interface - [V]: [(55, 1), (55, 3)]
2018-09-23 21:52:58,876 DEBUG - interface - [B]: [(56, 1), (56, 3)]
2018-09-23 21:52:58,876 DEBUG - interface - [N]: [(57, 1), (57, 3)]
2018-09-23 21:52:58,876 DEBUG - interface - [M]: [(58, 1), (58, 3)]
2018-09-23 21:52:58,876 DEBUG - interface - [<]: [(94, 0), (59, 1), (94, 2), (59, 3)]
2018-09-23 21:52:58,876 DEBUG - interface - [>]: [(60, 1), (94, 1), (60, 3), (94, 3)]
2018-09-23 21:52:58,876 DEBUG - interface - [?]: [(61, 1), (61, 3)]
2018-09-23 21:52:58,883 DEBUG - iomediator - Set modifier Key.CAPSLOCK to False
2018-09-23 21:52:58,883 DEBUG - iomediator - Set modifier Key.NUMLOCK to False
2018-09-23 21:52:58,884 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
2018-09-23 21:52:58,884 DEBUG - interface - __flushEvents: Entering event loop.
2018-09-23 21:52:58,884 INFO - iomediator - Created IoMediator instance, current interface is: <XRecordInterface(XInterface-thread, initial daemon)>
2018-09-23 21:52:58,887 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
[...]
2018-09-23 21:52:58,893 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
2018-09-23 21:52:58,894 INFO - interface - XRecord interface thread starting
2018-09-23 21:52:58,894 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
2018-09-23 21:52:58,894 INFO - service - Service now marked as running
2018-09-23 21:52:58,895 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
[...]
2018-09-23 21:52:58,895 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
2018-09-23 21:52:58,953 DEBUG - phrase-menu - Sorting phrase menu by usage count
2018-09-23 21:52:58,954 DEBUG - phrase-menu - Triggering menu item by position in list
2018-09-23 21:52:58,895 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
[...]
2018-09-23 21:52:58,895 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
2018-09-23 21:52:58,958 DEBUG - root - Created DBus service
2018-09-23 21:52:58,959 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
2018-09-23 21:52:58,959 INFO - root - Entering main()
[...]
2018-09-23 21:53:00,210 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>'
2018-09-23 21:53:00,211 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>'
[...]
2018-09-23 21:53:01,245 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>'
2018-09-23 21:53:01,245 DEBUG - interface - Grabbing hotkey: ['<ctrl>'] '<f7>'
2018-09-23 21:53:11,339 INFO - root - Shutting down
2018-09-23 21:53:11,339 INFO - service - Service shutting down
2018-09-23 21:53:11,339 DEBUG - iomediator - IoMediator shutting down
2018-09-23 21:53:11,340 DEBUG - interface - XInterfaceBase: Try to exit event thread.
2018-09-23 21:53:11,340 DEBUG - interface - XInterfaceBase: Event thread exit marker enqueued.
2018-09-23 21:53:11,340 DEBUG - interface - XInterfaceBase: self.shutdown set to True. This should stop the listener thread.
2018-09-23 21:53:12,016 DEBUG - interface - __flushEvents: Left event loop.
2018-09-23 21:53:12,018 DEBUG - iomediator - Waiting for IoMediator thread to end
2018-09-23 21:53:12,018 DEBUG - iomediator - IoMediator shutdown completed
2018-09-23 21:53:12,018 INFO - config-manager - Persisting configuration
2018-09-23 21:53:12,019 INFO - config-manager - Backing up existing config file
2018-09-23 21:53:12,020 DEBUG - inotify - Reporting IN_MODIFY event at /home/wviechtb/.config/autokey/autokey.json~
2018-09-23 21:53:12,021 DEBUG - inotify - Reporting IN_MODIFY event at /home/wviechtb/.config/autokey/autokey.json~
2018-09-23 21:53:12,022 DEBUG - inotify - Reporting IN_MODIFY event at /home/wviechtb/.config/autokey/autokey.json
2018-09-23 21:53:12,024 DEBUG - inotify - Reporting IN_MODIFY event at /home/wviechtb/.config/autokey/autokey.json
2018-09-23 21:53:12,024 INFO - config-manager - Finished persisting configuration - no errors
2018-09-23 21:53:12,025 DEBUG - service - Service shutdown completed.
2018-09-23 21:53:13,026 INFO - inotify - Shutting down file monitor
2018-09-23 21:53:13,026 DEBUG - root - All shutdown tasks complete... quitting
```

**NOTE**: Each `[...]` is thousands of repetitions of the same message (I cut them out because the log was more than 400kb).

## Version

AutoKey version: 0.95.3

Used GUI: Gtk

Installed via: pip3

Distro: Ubuntu 16.04

**Comment by luziferius** [comment]
I’ve seen this in quite some logs, but not that excessive. I am not quite sure what is the root cause.
Do you use window filters, especially negating ones (triggering on all windows except the specified one)?

The debug message isn’t that helpful, so would you mind altering it on your system?
You can find it here: https://github.com/autokey/autokey/blob/3e60a0161e3f3d0854dac5207de2f607aa20d7a3/lib/autokey/interface.py#L451-L455

So just open the interface.py file in an editor (maybe needs root rights, if you installed it globally), and replace that logger line in line 455 with this:
```
        logger.debug("Grabbing hotkey: %r %r for window: class=%r, title=%r",
                     modifiers, key, window.get_wm_class(), window.get_wm_name())
```
This adds the information for which windows the hotkey is grabbed.

The output then looks like this:
```
2018-09-24 13:32:01,022 DEBUG - interface - Grabbing hotkey: [<Key.CONTROL: '<ctrl>'>] 'ü' for window: class=('Navigator', 'Firefox'), title='Repeated "interface - Grabbing hotkey" causing high CPU load · Issue #193 · autokey/autokey - Mozilla Firefox'
```
_I think_, this relates to how _window_ is defined in X11 terms. As far as I know, each widget (like buttons) is it’s own window that belongs to some parent window. E.g. the save button in the toolbar is a window that belongs to the toolbar window, which belongs to the autokey window, which belongs to the root window (somewhat like the screen). 

Using filters forces autokey to go through this whole mess and grab the hotkey everywhere instead of just at the root level, causing that much CPU load.

**Comment by wviechtb** [comment]
Thanks for the response! Yes, I use lots of filters, but no negative ones. But I see the same problem even if I get rid of my phrases and scripts and just use the default ones. Output below as requested (with default phrases/scripts, not my own):

```
2018-09-24 15:53:02,860 INFO - root - Initialising application
2018-09-24 15:53:02,866 INFO - root - Initialise global hotkeys
2018-09-24 15:53:02,866 INFO - config-manager - Loading config from existing file: /home/wviechtb/.config/autokey/autokey.json
2018-09-24 15:53:02,866 DEBUG - config-manager - Loading folder at '/home/wviechtb/.config/autokey/data/Sample Scripts'
2018-09-24 15:53:02,867 DEBUG - config-manager - Loading folder at '/home/wviechtb/.config/autokey/data/My Phrases'
2018-09-24 15:53:02,868 INFO - config-manager - Configuration changed - rebuilding in-memory structures
2018-09-24 15:53:02,868 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey/data/Sample Scripts
2018-09-24 15:53:02,868 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey/data/My Phrases
2018-09-24 15:53:02,869 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey/data/My Phrases/Addresses
2018-09-24 15:53:02,869 INFO - config-manager - Successfully loaded configuration
2018-09-24 15:53:02,869 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey/data
2018-09-24 15:53:02,869 DEBUG - inotify - Adding watch for /home/wviechtb/.config/autokey
2018-09-24 15:53:02,869 DEBUG - config-manager - Global settings: {'showToolbar': False, 'undoUsingBackspace': True, 'promptToSave': False, 'interfaceType': 'XRecord', 'menuTakesFocus': False, 'triggerItemByInitial': False, 'notificationIcon': 'autokey-status', 'enableQT4Workaround': False, 'columnWidths': [150, 50, 100], 'windowDefaultSize': [2100, 1367], 'hPanePosition': 527, 'scriptGlobals': {}, 'serviceRunning': True, 'sortByUsageCount': True, 'isFirstRun': True, 'showTrayIcon': True, 'workAroundApps': '.*VirtualBox.*|krdc.Krdc'}
2018-09-24 15:53:02,869 INFO - service - Starting service
2018-09-24 15:53:02,889 DEBUG - interface - Modifier masks: {<Key.NUMLOCK: '<numlock>'>: 16, <Key.META: '<meta>'>: 8, <Key.SHIFT: '<shift>'>: 1, <Key.ALT_GR: '<alt_gr>'>: 128, <Key.ALT: '<alt>'>: 8, <Key.SUPER: '<super>'>: 64, <Key.CONTROL: '<ctrl>'>: 4, <Key.HYPER: '<hyper>'>: 64}
2018-09-24 15:53:02,912 DEBUG - interface - Alt-Grid: XK_Alt_R, 65514
2018-09-24 15:53:02,912 DEBUG - interface - <map object at 0x7f9220055a90>
2018-09-24 15:53:02,912 DEBUG - interface - X Server Keymap
2018-09-24 15:53:02,912 DEBUG - interface - [\]: [(51, 0), (51, 2)]
2018-09-24 15:53:02,912 DEBUG - interface - [|]: [(51, 1), (51, 3), (94, 4), (94, 6)]
2018-09-24 15:53:02,912 DEBUG - interface - [`]: [(49, 0), (49, 2)]
2018-09-24 15:53:02,912 DEBUG - interface - [1]: [(10, 0), (10, 2)]
2018-09-24 15:53:02,912 DEBUG - interface - [2]: [(11, 0), (11, 2)]
2018-09-24 15:53:02,912 DEBUG - interface - [3]: [(12, 0), (12, 2)]
2018-09-24 15:53:02,912 DEBUG - interface - [4]: [(13, 0), (13, 2)]
2018-09-24 15:53:02,912 DEBUG - interface - [5]: [(14, 0), (14, 2)]
2018-09-24 15:53:02,912 DEBUG - interface - [6]: [(15, 0), (15, 2)]
2018-09-24 15:53:02,913 DEBUG - interface - [7]: [(16, 0), (16, 2)]
2018-09-24 15:53:02,913 DEBUG - interface - [8]: [(17, 0), (17, 2)]
2018-09-24 15:53:02,913 DEBUG - interface - [9]: [(18, 0), (18, 2)]
2018-09-24 15:53:02,913 DEBUG - interface - [0]: [(19, 0), (19, 2)]
2018-09-24 15:53:02,913 DEBUG - interface - [-]: [(20, 0), (20, 2)]
2018-09-24 15:53:02,913 DEBUG - interface - [=]: [(21, 0), (21, 2)]
2018-09-24 15:53:02,913 DEBUG - interface - [~]: [(49, 1), (49, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [!]: [(10, 1), (10, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [@]: [(11, 1), (11, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [#]: [(12, 1), (12, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [$]: [(13, 1), (13, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [%]: [(14, 1), (14, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [^]: [(15, 1), (15, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [&]: [(16, 1), (16, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [*]: [(17, 1), (17, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [(]: [(187, 0), (18, 1), (187, 2), (18, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [)]: [(188, 0), (19, 1), (188, 2), (19, 3)]
2018-09-24 15:53:02,913 DEBUG - interface - [q]: [(24, 0), (24, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [w]: [(25, 0), (25, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [e]: [(26, 0), (26, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [r]: [(27, 0), (27, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [t]: [(28, 0), (28, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [y]: [(29, 0), (29, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [u]: [(30, 0), (30, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [i]: [(31, 0), (31, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [o]: [(32, 0), (32, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [p]: [(33, 0), (33, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [[]: [(34, 0), (34, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - []]: [(35, 0), (35, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [a]: [(38, 0), (38, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [s]: [(39, 0), (39, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [d]: [(40, 0), (40, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [f]: [(41, 0), (41, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [g]: [(42, 0), (42, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [h]: [(43, 0), (43, 2)]
2018-09-24 15:53:02,914 DEBUG - interface - [j]: [(44, 0), (44, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [k]: [(45, 0), (45, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [l]: [(46, 0), (46, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [;]: [(47, 0), (47, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [']: [(48, 0), (48, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [z]: [(52, 0), (52, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [x]: [(53, 0), (53, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [c]: [(54, 0), (54, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [v]: [(55, 0), (55, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [b]: [(56, 0), (56, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [n]: [(57, 0), (57, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [m]: [(58, 0), (58, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [,]: [(59, 0), (59, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [.]: [(60, 0), (60, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [/]: [(61, 0), (61, 2)]
2018-09-24 15:53:02,915 DEBUG - interface - [Q]: [(24, 1), (24, 3)]
2018-09-24 15:53:02,915 DEBUG - interface - [W]: [(25, 1), (25, 3)]
2018-09-24 15:53:02,915 DEBUG - interface - [E]: [(26, 1), (26, 3)]
2018-09-24 15:53:02,915 DEBUG - interface - [R]: [(27, 1), (27, 3)]
2018-09-24 15:53:02,915 DEBUG - interface - [T]: [(28, 1), (28, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [Y]: [(29, 1), (29, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [U]: [(30, 1), (30, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [I]: [(31, 1), (31, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [O]: [(32, 1), (32, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [P]: [(33, 1), (33, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [{]: [(34, 1), (34, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [}]: [(35, 1), (35, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [A]: [(38, 1), (38, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [S]: [(39, 1), (39, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [D]: [(40, 1), (40, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [F]: [(41, 1), (41, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [G]: [(42, 1), (42, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [H]: [(43, 1), (43, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [J]: [(44, 1), (44, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [K]: [(45, 1), (45, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [L]: [(46, 1), (46, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - [:]: [(47, 1), (47, 3)]
2018-09-24 15:53:02,916 DEBUG - interface - ["]: [(48, 1), (48, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [Z]: [(52, 1), (52, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [X]: [(53, 1), (53, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [C]: [(54, 1), (54, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [V]: [(55, 1), (55, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [B]: [(56, 1), (56, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [N]: [(57, 1), (57, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [M]: [(58, 1), (58, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [<]: [(94, 0), (59, 1), (94, 2), (59, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [>]: [(60, 1), (94, 1), (60, 3), (94, 3)]
2018-09-24 15:53:02,917 DEBUG - interface - [?]: [(61, 1), (61, 3)]
2018-09-24 15:53:02,917 DEBUG - iomediator - Set modifier Key.CAPSLOCK to False
2018-09-24 15:53:02,917 DEBUG - iomediator - Set modifier Key.NUMLOCK to False
2018-09-24 15:53:02,918 DEBUG - interface - __flushEvents: Entering event loop.
2018-09-24 15:53:02,919 INFO - iomediator - Created IoMediator instance, current interface is: <XRecordInterface(XInterface-thread, initial daemon)>
2018-09-24 15:53:02,934 INFO - interface - XRecord interface thread starting
2018-09-24 15:53:02,935 INFO - service - Service now marked as running
2018-09-24 15:53:02,981 DEBUG - phrase-menu - Sorting phrase menu by usage count
2018-09-24 15:53:02,982 DEBUG - phrase-menu - Triggering menu item by position in list
2018-09-24 15:53:02,985 DEBUG - root - Created DBus service
2018-09-24 15:53:02,986 INFO - root - Entering main()
2018-09-24 15:53:02,919 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title=None
2018-09-24 15:53:02,934 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('desktop_window', 'Nautilus'), title='Desktop'
2018-09-24 15:53:02,944 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('gnome-terminal-server', 'Gnome-terminal'), title='wviechtb@wvx1: ~'
2018-09-24 15:53:03,075 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('psensor', 'Psensor'), title='Psensor - Temperature Monitor'
2018-09-24 15:53:03,250 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='XdndCollectionWindowImp'
2018-09-24 15:53:03,252 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='unity-panel'
2018-09-24 15:53:03,253 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='unity-dash'
2018-09-24 15:53:03,254 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='Hud'
2018-09-24 15:53:03,255 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='unity-launcher'
2018-09-24 15:53:03,256 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('ibus-x11', 'Ibus-x11'), title='ibus-x11'
2018-09-24 15:53:03,256 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('bamfdaemon', 'Bamfdaemon'), title='bamfdaemon'
2018-09-24 15:53:03,256 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='ibus-xim'
2018-09-24 15:53:03,257 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('ibus-ui-gtk3', 'Ibus-ui-gtk3'), title='ibus-ui-gtk3'
2018-09-24 15:53:03,258 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-settings-daemon', 'Unity-settings-daemon'), title='unity-settings-daemon'
2018-09-24 15:53:03,258 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,259 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,260 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,261 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,262 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('indicator-printers-service', 'Indicator-printers-service'), title='indicator-printers-service'
2018-09-24 15:53:03,263 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('indicator-keyboard-service', 'Indicator-keyboard-service'), title='indicator-keyboard-service'
2018-09-24 15:53:03,266 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,267 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-fallback-mount-helper', 'Unity-fallback-mount-helper'), title='unity-fallback-mount-helper'
2018-09-24 15:53:03,268 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('compiz', 'Compiz'), title='compiz'
2018-09-24 15:53:03,270 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('nm-applet', 'Nm-applet'), title='nm-applet'
2018-09-24 15:53:03,272 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('indicator-multiload', 'Indicator-multiload'), title='indicator-multiload'
2018-09-24 15:53:03,276 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('polkit-gnome-authentication-agent-1', 'Polkit-gnome-authentication-agent-1'), title='polkit-gnome-authentication-agent-1'
2018-09-24 15:53:03,277 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('nautilus', 'Nautilus'), title='nautilus'
2018-09-24 15:53:03,278 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('redshift-gtk', ''), title='redshift-gtk'
2018-09-24 15:53:03,278 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('indicator-sysmonitor', ''), title='indicator-sysmonitor'
2018-09-24 15:53:03,279 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('compiz', 'Compiz'), title='compiz'
2018-09-24 15:53:03,280 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='compiz'
2018-09-24 15:53:03,281 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('ibus-ui-gtk', 'ibus-ui-gtk'), title='IBus Panel'
2018-09-24 15:53:03,282 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='nautilus'
2018-09-24 15:53:03,282 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='nautilus'
2018-09-24 15:53:03,283 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('gnome-software', 'Gnome-software'), title='gnome-software'
2018-09-24 15:53:03,283 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('psensor', 'Psensor'), title='psensor'
2018-09-24 15:53:03,284 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('update-notifier', 'Update-notifier'), title='update-notifier'
2018-09-24 15:53:03,284 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='update-notifier'
2018-09-24 15:53:03,285 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('gnome-terminal-server', 'Gnome-terminal-server'), title='Terminal'
2018-09-24 15:53:03,285 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=None, title='Terminal'
2018-09-24 15:53:03,286 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('notify-osd', 'Notify-osd'), title='notify-osd'
2018-09-24 15:53:03,287 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,288 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,290 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('psensor', 'Psensor'), title='psensor'
2018-09-24 15:53:03,291 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,292 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,293 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,293 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,294 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,295 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,298 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,298 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,299 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,300 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,301 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,302 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,306 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,311 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,312 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,313 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('nautilus', 'Nautilus'), title='nautilus'
2018-09-24 15:53:03,314 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('nautilus', 'Nautilus'), title='nautilus'
2018-09-24 15:53:03,315 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,316 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,316 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,317 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,318 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,319 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,319 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('gnome-terminal-server', 'Gnome-terminal'), title='Terminal'
2018-09-24 15:53:03,320 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('gnome-terminal-server', 'Gnome-terminal'), title='Terminal'
2018-09-24 15:53:03,321 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('nautilus', 'Nautilus'), title='nautilus'
2018-09-24 15:53:03,322 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,323 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('keynav', 'keynav'), title=None
2018-09-24 15:53:03,323 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,324 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,324 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,325 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,327 DEBUG - interface - Grabbing hotkey: ['<super>'] '<f12>' for window: class=('autokey-gtk', ''), title='autokey-gtk'
2018-09-24 15:53:03,330 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('desktop_window', 'Nautilus'), title='Desktop'
2018-09-24 15:53:03,346 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('gnome-terminal-server', 'Gnome-terminal'), title='wviechtb@wvx1: ~'
2018-09-24 15:53:03,458 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('psensor', 'Psensor'), title='Psensor - Temperature Monitor'
2018-09-24 15:53:03,624 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='XdndCollectionWindowImp'
2018-09-24 15:53:03,625 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='unity-panel'
2018-09-24 15:53:03,627 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='unity-dash'
2018-09-24 15:53:03,628 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='Hud'
2018-09-24 15:53:03,629 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='unity-launcher'
2018-09-24 15:53:03,630 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('ibus-x11', 'Ibus-x11'), title='ibus-x11'
2018-09-24 15:53:03,630 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('bamfdaemon', 'Bamfdaemon'), title='bamfdaemon'
2018-09-24 15:53:03,631 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='ibus-xim'
2018-09-24 15:53:03,631 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('ibus-ui-gtk3', 'Ibus-ui-gtk3'), title='ibus-ui-gtk3'
2018-09-24 15:53:03,632 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-settings-daemon', 'Unity-settings-daemon'), title='unity-settings-daemon'
2018-09-24 15:53:03,633 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,633 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,634 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,635 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,636 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('indicator-printers-service', 'Indicator-printers-service'), title='indicator-printers-service'
2018-09-24 15:53:03,637 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('indicator-keyboard-service', 'Indicator-keyboard-service'), title='indicator-keyboard-service'
2018-09-24 15:53:03,641 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,646 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-fallback-mount-helper', 'Unity-fallback-mount-helper'), title='unity-fallback-mount-helper'
2018-09-24 15:53:03,648 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('compiz', 'Compiz'), title='compiz'
2018-09-24 15:53:03,650 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('nm-applet', 'Nm-applet'), title='nm-applet'
2018-09-24 15:53:03,651 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('indicator-multiload', 'Indicator-multiload'), title='indicator-multiload'
2018-09-24 15:53:03,652 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('polkit-gnome-authentication-agent-1', 'Polkit-gnome-authentication-agent-1'), title='polkit-gnome-authentication-agent-1'
2018-09-24 15:53:03,653 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('nautilus', 'Nautilus'), title='nautilus'
2018-09-24 15:53:03,653 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('redshift-gtk', ''), title='redshift-gtk'
2018-09-24 15:53:03,654 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('indicator-sysmonitor', ''), title='indicator-sysmonitor'
2018-09-24 15:53:03,655 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('compiz', 'Compiz'), title='compiz'
2018-09-24 15:53:03,656 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='compiz'
2018-09-24 15:53:03,656 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('ibus-ui-gtk', 'ibus-ui-gtk'), title='IBus Panel'
2018-09-24 15:53:03,657 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='nautilus'
2018-09-24 15:53:03,658 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='nautilus'
2018-09-24 15:53:03,658 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('gnome-software', 'Gnome-software'), title='gnome-software'
2018-09-24 15:53:03,659 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('psensor', 'Psensor'), title='psensor'
2018-09-24 15:53:03,659 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('update-notifier', 'Update-notifier'), title='update-notifier'
2018-09-24 15:53:03,660 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='update-notifier'
2018-09-24 15:53:03,660 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('gnome-terminal-server', 'Gnome-terminal-server'), title='Terminal'
2018-09-24 15:53:03,661 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=None, title='Terminal'
2018-09-24 15:53:03,661 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('notify-osd', 'Notify-osd'), title='notify-osd'
2018-09-24 15:53:03,662 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,663 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,665 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('psensor', 'Psensor'), title='psensor'
2018-09-24 15:53:03,666 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,667 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,667 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,668 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,669 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,670 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,672 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,673 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,674 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,675 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,677 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,679 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,681 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,684 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,686 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,687 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('nautilus', 'Nautilus'), title='nautilus'
2018-09-24 15:53:03,688 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('nautilus', 'Nautilus'), title='nautilus'
2018-09-24 15:53:03,689 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,690 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,691 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,692 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,692 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,693 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('hud-service', 'Hud-service'), title='hud-service'
2018-09-24 15:53:03,694 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('gnome-terminal-server', 'Gnome-terminal'), title='Terminal'
2018-09-24 15:53:03,695 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('gnome-terminal-server', 'Gnome-terminal'), title='Terminal'
2018-09-24 15:53:03,696 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('nautilus', 'Nautilus'), title='nautilus'
2018-09-24 15:53:03,696 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,697 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('keynav', 'keynav'), title=None
2018-09-24 15:53:03,697 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,698 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,699 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,700 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('unity-panel-service', 'Unity-panel-service'), title='unity-panel-service'
2018-09-24 15:53:03,702 DEBUG - interface - Grabbing hotkey: ['<super>'] '<delete>' for window: class=('autokey-gtk', ''), title='autokey-gtk'
2018-09-24 15:53:03,702 DEBUG - interface - Grabbing hotkey: ['<ctrl>'] '<f7>' for window: class=None, title=None
2018-09-24 15:53:10,186 INFO - root - Shutting down
2018-09-24 15:53:10,186 INFO - service - Service shutting down
2018-09-24 15:53:10,186 DEBUG - iomediator - IoMediator shutting down
2018-09-24 15:53:10,186 DEBUG - interface - XInterfaceBase: Try to exit event thread.
2018-09-24 15:53:10,187 DEBUG - interface - XInterfaceBase: Event thread exit marker enqueued.
2018-09-24 15:53:10,187 DEBUG - interface - XInterfaceBase: self.shutdown set to True. This should stop the listener thread.
2018-09-24 15:53:10,983 DEBUG - interface - __flushEvents: Left event loop.
2018-09-24 15:53:10,986 DEBUG - iomediator - Waiting for IoMediator thread to end
2018-09-24 15:53:10,986 DEBUG - iomediator - IoMediator shutdown completed
2018-09-24 15:53:10,987 INFO - config-manager - Persisting configuration
2018-09-24 15:53:10,987 INFO - config-manager - Backing up existing config file
2018-09-24 15:53:10,989 DEBUG - inotify - Reporting IN_MODIFY event at /home/wviechtb/.config/autokey/autokey.json~
2018-09-24 15:53:10,992 DEBUG - inotify - Reporting IN_MODIFY event at /home/wviechtb/.config/autokey/autokey.json
2018-09-24 15:53:10,993 DEBUG - inotify - Reporting IN_MODIFY event at /home/wviechtb/.config/autokey/autokey.json
2018-09-24 15:53:10,994 INFO - config-manager - Finished persisting configuration - no errors
2018-09-24 15:53:10,995 DEBUG - service - Service shutdown completed.
2018-09-24 15:53:11,995 INFO - inotify - Shutting down file monitor
2018-09-24 15:53:12,002 DEBUG - root - All shutdown tasks complete... quitting
```

**NOTE:** I removed lots of repeated rows where `class=None, title=None` (there were many of them, interspersed with the rest), since these are not informative. The full log had 1628 rows.

**Comment by wviechtb** [comment]
Is there anything else I can provide that could help to diagnose the source of this issue? I use Autokey extensively, but the use of certain hotkeys leads to a

`DEBUG - interface - Recorded keymap change event`

event in which case I first get tons of repeated

`DEBUG - interface - Ungrabbing hotkey: [...]`

events followed by tons of repeated

`DEBUG - interface - Grabbing hotkey:`

events. In these moments, CPU load jumps close to 100% for a few seconds and Autokey becomes unresponsive.

As long as there is no `Recorded keymap change event` event, everything runs super smooth. So, even if there is no immediate solution for preventing the repeated grabbing of the same keys, would there be a way to suppress triggering of the `keymap change event`? That might prevent the excessive intermittent CPU load. Just an idea.

**Comment by luziferius** [comment]
I haven’t looked at the interface stuff and reverse-engineered how exactly it works.

Let’s try to do your suggestion. I looked around where the keymap change event is processed. Looks like it is here:
https://github.com/autokey/autokey/blob/3e60a0161e3f3d0854dac5207de2f607aa20d7a3/lib/autokey/interface.py#L239-L251
You can try to comment out or remove line 251 in `interface.py`. This will cause the code to be executed only once at start-up.
I don’t know if this will have any implications or cause issues.

**Comment by wviechtb** [comment]
I tried commenting out 251, but that did not help (although it seems like it should ...). Then I commented out 244 and 245 and that stopped the remapping. Maybe a bit drastic, but so far this has stopped the CPU from ramping up. I'll need to do some more testing, but so far things seem to work just fine .. ¯\\_(ツ)_/¯

**Comment by octodapple** [comment]
I confirm this bug, exactly as described by @wviechtb, in versions 0.90.4 and 0.95.4, in Ubuntu 18.04.1. The high cpu, and 1000s of 'grabbing' debug messages happens on startup, but is also triggered by changing the system keymap (in Ubuntu settings, region & language, input sources).
Worse, it triggers high cpu for about 3-5 seconds on every generated keypress from xdotool, xte and xvkbd, for example
```
xdotool key a
```
will trigger it, whether xdotool is called directly from the terminal or from xbindkeys. It is not triggered by an xdotool mouse button click, eg
```
xdotool click 1
```
So it appears that autokey is detecting that the `xdotool key` command has changed the keymap.
The `Recorded keymap change event` appears in the autokey log every time xdotool key is triggered.
@wviechtb, you can check if your hotkeys that trigger this contain xdotool or similar.

Commenting out line 214 in `interface.py`, in the `XInterfaceBase init__()` function, which calls on_keys_changed(), mentioned above by @luziferius, fixes it or greatly reduces the effect.
```
self.keyMap.connect("keys-changed", self.on_keys_changed)
```
This will be a problem if you are regularly switching between keymaps. Otherwise you can try restarting autokey after changing your keymap. Everything else I've tried works fine with this.

The 1000s of debug lines are coming from looping through `children` in `__recurseTree()`. If you debug `len(children)` it corresponds to the number shown by running this in your shell:
```
xwininfo -root -children | grep children
```
which is a list of window ids for all windows, tabs and every window manager control and widget. If you have say 10 programs open, there will be >100 items in this list.
`__recurseTree()` then runs a nested loop through every hotkey defined in autokey and runs `__grabHotkey()` for each one, which is where the `Grabbing hotkey` log messages come from.
On startup, autokey is running this on the two hotkeys defined in autokey's preferences - special hotkeys page. You can prevent that by clearing the two hotkeys there.
Closing as many windows + tabs as possible, and having as few autokey hotkeys defined as possible will also reduce the effect.

I've checked the xdotool documentation, there is no log or verbose option etc, or any mention of keymaps changing.
I've tried to fix it on the xdotool end with --clearmodifiers, --delay, sleep, windowfocus to specific window, nothing makes a difference.

I've also noticed that with autokey not running, xdotool and xte trigger high cpu in the xorg process, for about one second. And changing the system keymap also causes it.

**Comment by josephj11** [comment]
This is starting to sound like a basic design issue. AutoKey might need to grab hotkeys, etc. at a much lower level in X - before all the functionality get delegated to all the various "windows". However, doing so doesn't really sound like a very good idea given the complexity of X and things like Wayland which may replace it.

I don't know if this is relevant, but I have a KDE startup script that does a bunch of stuff with xdotool  to start apps and rearrange windows. I get the application window count for the particular app I'm working with at the moment. That count is often much larger than 1 or 2 that I originally expected. Once I have the window count, I go through the list (stack) of windows xdotool returns in reverse order asking what desktop each window is on. Doing this, the "real" windows show up first (at the end of the list) and the rest return a desktop of -1. I ignore all of the ones that return -1.

I don't know if any of this is relevant here, but, if it is, maybe there's a way for AutoKey to ignore a bunch of "windows" and cut down its processing load.

**Comment by wviechtb** [comment]
@octodapple Thank you for this extensive analysis. Indeed, just running `xdotool key a` triggers the high CPU load. I commented out 244 and 245 and that pretty much eliminated the issue for me and things run very smoothly now. I also tried your suggestion and instead commented out 214 and that also works. Of course, these are not proper fixes, but makes Autokey usable for me.

**Comment by octodapple** [comment]
I posted an [issue](https://github.com/jordansissel/xdotool/issues/231) over on xdotool, the main dev there says it sounds like an xdotool bug.

@josephj11, regarding going through the window stack, you could try using this instead:
```
wmctrl -l | grep -i <app name>
```
which only lists one id per open app. Then using the window id in hex, including the `0x`, pass that to `xdotool windowmove` etc.

Going back the autokey excessive window grabbing, looking at the output of
```
xwininfo -root -children | grep -E 'has no name|$'
```
shows a lot of unnamed windows. I tried eliminating these in autokey in [interface.py](https://github.com/autokey/autokey/blob/master/lib/autokey/interface.py) where these lines appear:
```
for window in children:
...
    title = self.get_window_title(window, False)
```
by adding:
```
if title == '': continue
```
However this broke some autokey hotkeys in some apps. Going on your suggestion of reversing the window stack, looking for the 'real' window, I also tried 
```
children = list(reversed(children))
```
and then adding a `break` statement in various place, but it again broke some hotkeys.

In any case, this high cpu in autokey should only occur on first load, or when the keymap is changed.
So it looks like the best option is to wait for a possible fix in xdotool, and in the meantime, use autokey with this line or some related line commented out, as mentioned above.
```
self.keyMap.connect("keys-changed", self.on_keys_changed)
```

**Comment by chrislee35** [comment]
This looks like the same mechanism fixed by #1235 and #1236 (both merged into develop): on_keys_changed() was doing a full ungrab/re-read/regrab of every hotkey on any MappingNotify event, even when the mapping hadn't actually changed. That matches what was traced here (a single event costing 8.98s and 10,666 XGrabKey round-trips), and #1236 also covers the case where AutoKey's own keyboard remapping was triggering the same regrab storm against itself.

Thanks to everyone who dug into the root cause in this thread, that diagnosis (tracing it to keys-changed and __recurseTree) lines up closely with what #1235/#1236 fixed.

If anyone still sees the high CPU load on current develop, please reopen or comment with details. Otherwise this looks closeable.

**Comment by josephj11** [comment]
Fixed by #1235 and #1236
